### PR TITLE
tikv/gc_worker: Add support for distributed GC

### PR DESCRIPTION
**Pull request by MyonKeminta**
Thank you for working on TiDB! Please read TiDB's [CONTRIBUTING](https://github.com/pingcap/tidb/blob/master/CONTRIBUTING.md) document **BEFORE** filing this PR.

## What have you changed? (mandatory)

Currently GC is done in 3 steps:
1. Resolve Locks
2. Delete Ranges
3. Do GC

In this PR, the third phase of GC does not do real GC one by one, but only send the new safe point to PD. Then TiKV can get the safe point from PD and do GC by themselves.

**IMPORTANT:** This PR causes incompatibility with old versions of PD and TiKV. 

For PD, this commit is required: 
* https://github.com/pingcap/pd/commit/629e0df8358c00cb89f030fee0c72465487b4345

For TiKV, this PR is required:
* https://github.com/tikv/tikv/pull/3179

If the TiKV or PD is too old, TiKV will never GC.

## What are the type of the changes (mandatory)?

- New feature (non-breaking change which adds functionality)

## How has this PR been tested (mandatory)?

By manually testing in a cluster.

## Does this PR affect documentation (docs/docs-cn) update? (optional)

Working in progress.

## Refer to a related PR or issue link (optional)

* https://github.com/pingcap/pd/pull/1108
* https://github.com/pingcap/tikv/pull/3179

## Add a few positive/negative examples (optional)

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/6833)

### CODE REVIEW

**Comment by zhexuany** [comment]
@MyonKeminta I saw you update `vendor`. In my opnion, it is better to update vendor first and then rebase against such updated master. In this way, you can benefit two things:
1. reduce code change size; it make PR easier to be reviewed 
2. avode the potential conflict with `Gopkg.lock`. 

Note: `Gopkg.lock` is generated by `dep`, we should be better not to touch it.

File: store/tikv/gcworker/gc_worker.go [hunk 1]
```diff
@@ -642,6 +643,26 @@ func (w *GCWorker) resolveLocks(ctx context.Context, safePoint uint64) error {
...
+		newSafePoint, err = w.pdClient.UpdateGCSafePoint(ctx, safePoint)
+		if err == nil {
+			break
+		}
+	}
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if newSafePoint != safePoint {
+		log.Warnf("[gc worker] pd rejected our safe point %v, because pd has greater safe point %v", safePoint, newSafePoint)
```
Comment (zhangjinpeng1987) [inline store/tikv/gcworker/gc_worker.go]: Is it possible that pd reject the update safepoint request and pd's safepoint is less than tidb's safepoint?

File: store/tikv/gcworker/gc_worker.go [hunk 2]
```diff
@@ -642,6 +643,26 @@ func (w *GCWorker) resolveLocks(ctx context.Context, safePoint uint64) error {
...
+		// TODO: If unfortunately this happened, should we do something else?
```
Comment (zhangjinpeng1987) [inline store/tikv/gcworker/gc_worker.go]: Add a metric for this case, and trigger an alert from the metric.

File: store/tikv/gcworker/gc_worker.go [hunk 3]
```diff
@@ -122,6 +122,11 @@ const (
 	gcEnableValue        = "true"
 	gcDisableValue       = "false"
 	gcDefaultEnableValue = true
+
+	gcModeKey         = "tikv_gc_mode"
+	gcModeLegacy      = "legacy"
```
Comment (zhangjinpeng1987) [inline store/tikv/gcworker/gc_worker.go]: Central is better, legacy doesn't have a exact meaning.

File: store/tikv/gcworker/gc_worker.go [hunk 4]
```diff
@@ -400,14 +406,35 @@ func (w *GCWorker) runGCJob(ctx context.Context, safePoint uint64) {
...
 		return
 	}
-	err = w.doGC(ctx, safePoint)
+
+	useDistributedGC, err := w.checkUseDistributedGC()
 	if err != nil {
-		log.Errorf("[gc worker] %s do GC returns an error %v", w.uuid, errors.ErrorStack(err))
-		w.gcIsRunning = false
-		metrics.GCJobFailureCounter.WithLabelValues("gc").Inc()
+		log.Errorf("[gc worker] %s failed to load gc mode: %v", w.uuid, errors.ErrorStack(err))
```
Comment (zhangjinpeng1987) [inline store/tikv/gcworker/gc_worker.go]: If check gc mode failed, should we fallback to old mode?

Comment (MyonKeminta) [reply store/tikv/gcworker/gc_worker.go]: Emmm, good idea

Comment (MyonKeminta) [reply store/tikv/gcworker/gc_worker.go]: So should the default mode be the old one?

**Comment by MyonKeminta** [comment]
@zhangjinpeng1987 It seems the word `centered` is more accurate than `central`?

**Comment by MyonKeminta** [comment]
@zhangjinpeng1987 @Connor1996 PTAL again

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**